staging/vhost-device-video/src/vhu_video.rs

```diff
@@ -0,0 +1,264 @@
+    #[error("Failed to consume new epoll event: {0}")]
+    EpollWait(std::io::Error),
+    #[error("Failed to de-register fd from epoll: {0}")]
+    EpollRemove(std::io::Error),
+    #[error("No memory configured")]
+    NoMemoryConfigured,
+    #[error("Unable to create thread pool: {0}")]
+    CreateThreadPool(std::io::Error),
+}
+
+impl convert::From<VuVideoError> for io::Error {
+    fn from(e: VuVideoError) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+#[derive(ValueEnum, Debug, Default, Clone, Eq, PartialEq)]
+pub(crate) enum BackendType {
+    #[default]
+    Null,
+}
+
+/// Virtio Video Configuration
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
+#[repr(C)]
+pub(crate) struct VirtioVideoConfig {
+    version: Le32,
+    max_caps_length: Le32,
+    max_resp_length: Le32,
+    device_name: [u8; VIRTIO_V4L2_CARD_NAME_LEN],
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioVideoConfig {}
+
+pub(crate) struct VuVideoBackend {
+    config: VirtioVideoConfig,
+    pub threads: Vec<Mutex<VhostUserVideoThread>>,
+    pub exit_event: EventFd,
+}
+
+impl VuVideoBackend {
+    /// Create a new virtio video device for /dev/video<num>.
+    pub fn new(video_path: &Path, video_backend: BackendType) -> Result<Self> {
+        let backend = Arc::new(RwLock::new(video_backends::alloc_video_backend(
+            video_backend,
+            video_path,
+        )?));
+        Ok(Self {
+            config: VirtioVideoConfig {
+                version: 0.into(),
+                max_caps_length: MAX_CAPS_LEN.into(),
+                max_resp_length: MAX_RESP_LEN.into(),
+                device_name: [0; 32],
+            },
+            threads: vec![Mutex::new(VhostUserVideoThread::new(backend.clone())?)],
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| VuVideoError::EventFdError)?,
+        })
+    }
+}
+
+/// VhostUserBackend trait methods
+impl VhostUserBackendMut<VringRwLock, ()> for VuVideoBackend {
+    fn num_queues(&self) -> usize {
+        NUM_QUEUES
+    }
+
+    fn max_queue_size(&self) -> usize {
+        QUEUE_SIZE
+    }
+
+    fn features(&self) -> u64 {
+        debug!("Get features");
+        1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_RING_F_INDIRECT_DESC
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | 1 << VIRTIO_VIDEO_F_RESOURCE_GUEST_PAGES
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        debug!("Get protocol features");
+        VhostUserProtocolFeatures::MQ
+            | VhostUserProtocolFeatures::CONFIG
+            | VhostUserProtocolFeatures::LOG_SHMFD
+            | VhostUserProtocolFeatures::HOST_NOTIFIER
+            | VhostUserProtocolFeatures::SLAVE_REQ
+            | VhostUserProtocolFeatures::SLAVE_SEND_FD
```

Comment:
Wondering if you need all of these, specially the SLAVE ones.

Reply:
Probably not, when I wrote these I looked into what qemu backends were setting, but I agree in that most are unneeded, I will remove those.